project_euler/problem_112/sol1.py

```diff
@@ -0,0 +1,89 @@
+"""
+Problem 112: https://projecteuler.net/problem=112
+
+Working from left-to-right if no digit is exceeded by the digit to its left it is
+called an increasing number; for example, 134468.
+Similarly if no digit is exceeded by the digit to its right it is called a decreasing
+number; for example, 66420.
+We shall call a positive integer that is neither increasing nor decreasing a "bouncy"
+number, for example, 155349.
+Clearly there cannot be any bouncy numbers below one-hundred, but just over half of
+the numbers below one-thousand (525) are bouncy. In fact, the least number for which
+the proportion of bouncy numbers first reaches 50% is 538.
+Surprisingly, bouncy numbers become more and more common and by the time we reach
+21780 the proportion of bouncy numbers is equal to 90%.
+
+Find the least number for which the proportion of bouncy numbers is exactly 99%.
+"""
+
+
+def check_bouncy(n: int) -> bool:
+    """
+    Returns True if number is bouncy, False otherwise
+    >>> check_bouncy(6789)
+    False
+    >>> check_bouncy(-12345)
+    False
+    >>> check_bouncy(0)
+    False
+    >>> check_bouncy(6.74)
+    Traceback (most recent call last):
+        ...
+    ValueError: check_bouncy() accepts only integer arguments
+    >>> check_bouncy(132475)
+    True
+    >>> check_bouncy(34)
+    False
+    >>> check_bouncy(341)
+    True
+    >>> check_bouncy(47)
+    False
+    >>> check_bouncy(-12.54)
+    Traceback (most recent call last):
+        ...
+    ValueError: check_bouncy() accepts only integer arguments
+    >>> check_bouncy(-6548)
+    True
+    """
+    if not isinstance(n, int):
+        raise ValueError("check_bouncy() accepts only integer arguments")
+    return "".join(sorted(str(n))) != str(n) and "".join(sorted(str(n)))[::-1] != str(n)

```

Comment:
```suggestion
    chars = str(abs(n))
    return not all(chars[i] <= chars[i + 1] for i in range(0, len(chars) - 1)) and \
        not all(chars[i] >= chars[i + 1] for i in range(0, len(chars) - 1))
```

Reply:
sorting is slow.

Reply:
Not true: (`cb1` is Kush, `cb2` is shellhub)
Note: this is not a competition, just a note that `sorting` itself could be slow but we have to look at the whole function.
```python
>>> timeit("cb1(47)", globals=globals())
0.7520698490000086
>>> timeit("cb2(47)", globals=globals())
0.9376681609999906
>>> timeit("cb1(341)", globals=globals())
1.5993510109999818
>>> timeit("cb2(341)", globals=globals())
1.8564014479999855
>>> timeit("cb1(6789)", globals=globals())
0.7960041720000106
>>> timeit("cb2(6789)", globals=globals())
1.2885217760000103
>>> timeit("cb1(-12345)", globals=globals())
0.8492059290000213
>>> timeit("cb2(-12345)", globals=globals())
1.3374689419999868
>>> timeit("cb1(-6548)", globals=globals())
1.7080608680000182
>>> timeit("cb2(-6548)", globals=globals())
2.0229644939999503
```

Reply:
Nice test.